src/inmanta/ast/statements/generator.py

```diff
@@ -268,7 +267,151 @@ def nested_blocks(self) -> Iterator["BasicBlock"]:
+        """
+        Resume when the iterable is fully ready for execution. Populates the helper's result variable and signals that no more
+        values will be sent.
+        """
+        # fetch helper, passed via the requires object
+        gradual_helper: object = requires[self].get_value()
+        assert isinstance(gradual_helper, ListComprehensionGradual)
+
+        iterable: object = self.iterable.execute({k: v.get_value() for k, v in requires.items()}, resolver, queue)
+        if isinstance(iterable, Unknown):
+            return Unknown(self)
+        if not isinstance(iterable, list):
+            # TODO: allow dict?
+            raise TypingException(
+                self, f"A list comprehension can only be applied to lists and relations, got {type(iterable)}"
+            )
+        if gradual_helper.nb_results_received == 0:
+            # non-gradual mode: pass results now
+            for item in iterable:
+                # TODO: should this be self.iterable.location?
+                gradual_helper.receive_result(item, self.location)
+        # indicate to helper that we're done
+        gradual_helper.done()
+
+    def execute(self, requires: dict[object, object], resolver: Resolver, queue: QueueScheduler) -> object:
+        return requires[self]
+
+    # TODO: execute_direct?
+    # TODO: repr/str
+
+    # TODO
+
+
+# TODO: name + invariants in docstring
+class ListComprehensionGradual(ResultCollector[object]):
+    """
+    Result collector for gradual execution of the list comprehension statement. When it receives a result, it sets up gradual
+    execution for the value expression, with the lhs as final result collector. Additionally, it collects all final results
+    in its own result variable for non-gradual execution.
+    """
+
+    def __init__(
+        self,
+        statement: ListComprehension,
+        resolver: Resolver,
+        queue: QueueScheduler,
+        lhs: Optional[ResultCollector[object]] = None,
+    ) -> None:
+        self.statement: ListComprehension = statement
+        self.resolver: Resolver = resolver
+        self.queue: QueueScheduler = queue
+        self.lhs: Optional[ResultCollector[object]] = None
+        self.nb_results_received: int = 0
+        self.result: FixedCountVariable = FixedCountVariable()
+
+    def receive_result(self, value: object, location: Location) -> bool:
+        # TODO: keep track of seen to avoid duplicate work? Be careful not to introduce the same bug as with the for loop
+        value_wrapper: ResultVariable[object] = ResultVariable()
+        value_wrapper.set_value(value, location)
+        value_resolver: VariableResolver = VariableResolver(
+            parent=self.resolver,
+            name=str(self.statement.loop_var),
+            variable=value_wrapper,
+        )
+
+        # execute the value expression
+        requires: dict[object, VariableABC] = (
+            self.statement.value_expression.requires_emit(value_resolver, self.queue)
```

Comment:
I think it is safe to just call requires_emit_gradual with a None value a last argument?